apps/site/assets/ts/trip-planner/components/TripPlannerPage.tsx

```diff
@@ -0,0 +1,114 @@
+import React, { Dispatch, ReactElement, useReducer, useEffect } from "react";
+import LeafletMap from "../../leaflet/components/Map";
+import {
+  MapData,
+  MapMarker as Marker
+} from "../../leaflet/components/__mapdata";
+import getBounds from "../../leaflet/bounds";
+
+interface Props {
+  mapData: MapData;
+}
+
+export interface State {
+  markers: Marker[];
+}
+
+interface UpdateAction {
+  type: "UPDATE_MARKER";
+  payload: {
+    marker: Marker;
+  };
+}
+
+interface RemoveAction {
+  type: "REMOVE_MARKER";
+  payload: {
+    id: string;
+  };
+}
+
+export type Action = UpdateAction | RemoveAction;
+
+export type RemoveEvent = Event & { detail: { label: string } };
+export type UpdateEvent = Event & {
+  detail: {
+    latitude: number;
+    longitude: number;
+    label: string;
+    title: string;
+  };
+};
+
+export const onRemove = (
+  { detail: { label } }: RemoveEvent,
+  dispatch: Dispatch<Action>
+): void => {
+  dispatch({ type: "REMOVE_MARKER", payload: { id: label } });
+};
+
+export const onUpdate = (
+  { detail }: UpdateEvent,
+  dispatch: Dispatch<Action>
+): void => {
+  const marker: Marker = {
+    icon: `map-pin-${detail.label.toLowerCase()}`,
+    id: detail.label,
+    latitude: detail.latitude,
+    longitude: detail.longitude,
+    rotation_angle: 0, // eslint-disable-line @typescript-eslint/camelcase
+    tooltip: <div>{detail.title}</div>
+  };
+
+  dispatch({ type: "UPDATE_MARKER", payload: { marker } });
+};
+
+const removeMarker = (markers: Marker[], id: string | null): Marker[] =>
+  markers.reduce(
+    (acc: Marker[], m: Marker) => (m.id === id ? acc : acc.concat(m)),
+    []
+  );
+
+export const reducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case "UPDATE_MARKER":
+      return {
+        markers: removeMarker(state.markers, action.payload.marker.id).concat(
+          action.payload.marker
+        )
+      };
+
+    case "REMOVE_MARKER":
+      return { markers: removeMarker(state.markers, action.payload.id) };
+
+    default:
+      // @ts-ignore
+      throw new Error(`unexpected action type: ${action.type}`);
+  }
+};
+
+const TripPlannerPage = ({ mapData }: Props): ReactElement<HTMLElement> => {
+  const [state, dispatch] = useReducer(reducer, { markers: mapData.markers });
+  useEffect(() => {
+    const onRemoveCb = (ev: Event): void =>
+      onRemove(ev as RemoveEvent, dispatch);
+    const onUpdateCb = (ev: Event): void =>
+      onUpdate(ev as UpdateEvent, dispatch);
+
+    document.removeEventListener("trip-plan:remove-marker", onRemoveCb);
+    document.addEventListener("trip-plan:remove-marker", onRemoveCb);
+
+    document.removeEventListener("trip-plan:update-marker", onUpdateCb);
+    document.addEventListener("trip-plan:update-marker", onUpdateCb);
```

Comment:
Are you able to move the `document.removeEventListener`s to the returned cleanup function from useEffect?

as in the last line of useEffect being
```
return () => { 
document.removeEventListener("trip-plan:remove-marker", onRemoveCb); 
document.removeEventListener("trip-plan:update-marker", onUpdateCb);
};
```

Reply:
Do you mean remove them from the main function, or have them in both places? my understanding is that when called with only a function, `useEffect` will run on every render. We don't have anything meaningful to pass as a memo argument, so i added the `removeEventListeners` inside it so it doesn't add a new listener every render.

Reply:
I'm referring to returning a function at the end of the useEffect function, here's the documentation:
https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

It's a convention to do clean-up by returning a function, and the convention will do a cleanup of the listeners on the unmount.